src/gpu_sysman.c

```diff
@@ -0,0 +1,1939 @@
+  bool temp;
+  bool throttle;
+} gpu_disable_t;
+
+/* handles for the GPU devices discovered by Sysman library */
+typedef struct {
+  char *pci_bdf;
+  char *dev_file;
+  /* number of types for metrics without allocs */
+  uint32_t ras_count;
+  uint32_t temp_count;
+  /* number of types for each counter metric */
+  uint32_t engine_count;
+  uint32_t membw_count;
+  uint32_t power_count;
+  uint32_t throttle_count;
+  /* number of types for each sampled metric */
+  uint32_t frequency_count;
+  uint32_t memory_count;
+  /* previous values for counters */
+  zes_engine_stats_t *engine;
+  zes_mem_bandwidth_t *membw;
+  zes_power_energy_counter_t *power;
+  zes_freq_throttle_time_t *throttle;
+  /* types * samples sized array of values, used for aggregate outputs */
+  zes_freq_state_t **frequency;
+  zes_mem_state_t **memory;
+  /* GPU  specific disable flags */
+  gpu_disable_t disabled;
+  zes_device_handle_t handle;
+  /* report counter */
+  uint64_t counter;
+} gpu_device_t;
+
+typedef enum {
+  OUTPUT_UNSET = 0,
+  OUTPUT_RAW,
+  OUTPUT_DERIVED,
+  OUTPUT_BOTH, /* 3 = 1 | 2 mask */
+  OUTPUT_TYPES
+} output_t;
+
+static const char *metrics_output[OUTPUT_TYPES] = {"unset", "raw", "derived",
+                                                   "both"};
+
+static gpu_device_t *gpus;
+static uint32_t gpu_count;
+static struct {
+  bool gpuinfo;
+  gpu_disable_t disabled;
+  output_t output;
+  uint32_t samples;
+} config;
+
+/* Sysman GPU plugin config options (defines to ease catching typos) */
+#define KEY_DISABLE_ENGINE "DisableEngine"
+#define KEY_DISABLE_ENGINE_SINGLE "DisableEngineSingle"
+#define KEY_DISABLE_FREQ "DisableFrequency"
+#define KEY_DISABLE_MEM "DisableMemory"
+#define KEY_DISABLE_MEMBW "DisableMemoryBandwidth"
+#define KEY_DISABLE_POWER "DisablePower"
+#define KEY_DISABLE_RAS "DisableErrors"
+#define KEY_DISABLE_RAS_SEPARATE "DisableSeparateErrors"
+#define KEY_DISABLE_TEMP "DisableTemperature"
+#define KEY_DISABLE_THROTTLE "DisableThrottleTime"
+
+#define KEY_METRICS_OUTPUT "MetricsOutput"
+#define KEY_LOG_GPU_INFO "LogGpuInfo"
+#define KEY_SAMPLES "Samples"
+#define MAX_SAMPLES 64
+
+/* Free array of arrays allocated with gpu_subarray_realloc().
+ *
+ * config.samples must not have changed since allocation, because
+ * that determines the number of allocated subarrays
+ */
+static bool gpu_subarray_free(void **mem) {
+  uint32_t i;
+  if (!mem) {
+    return false;
+  }
+  for (i = 0; i < config.samples; i++) {
+    free(mem[i]);
+    mem[i] = NULL;
+  }
+  free(mem);
+  return true;
+}
+
+/* Allocate 'config.samples' sized array of 'count' sized arrays having 'size'
+ * sized items.  If given array is already allocated, it and its subarrays
+ * is freed first
+ */
+static void **gpu_subarray_realloc(void **mem, int count, int size) {
+  uint32_t i;
+  gpu_subarray_free(mem);
+  mem = malloc(config.samples * sizeof(void *));
+  assert(mem);
+  for (i = 0; i < config.samples; i++) {
+    mem[i] = calloc(count, size);
```

Comment:
same here?

Reply:
I do not see calloc wrapper in common.c, only the smalloc wrapper for malloc?

Reply:
there is only the smalloc wrapper, no specific form for calloc.

Reply:
Would you like me to include also commit adding scalloc() wrapper to common.c?